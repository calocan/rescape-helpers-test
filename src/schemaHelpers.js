/**
 * Created by Andy Likuski on 2019.08.08
 * Copyright (c) 2019 Andy Likuski
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import * as R from 'ramda';
import {remoteLinkedSchemaTask, remoteSchemaTask} from '@rescapes/apollo';


/**
 * Creates a remote schema and adds resolvers with the given unary function expecting the schema
 * @param config
 * @param addResolvers
 * @returns {Task}
 */
export const resolvedRemoteSchemaTask = config => remoteLinkedSchemaTask(config);
/**
 * Returns a
 * @param config
 * @returns {Task<Object>} Resolves to an apolloClient
 */
export const remoteApolloClientTask = config => R.map(
  ({apolloClient}) => apolloClient,
  remoteSchemaTask(config)
);
