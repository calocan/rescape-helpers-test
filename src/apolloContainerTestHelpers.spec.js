import {
  apolloContainerTests, makeApolloTestPropsTaskFunction,
  propsFromParentPropsTask
} from './apolloContainerTestHelpers';
import {gql} from 'apollo-client-preset';
import * as R from 'ramda';
import {connect} from 'react-redux';
import {graphql} from 'react-apollo';
import {Component} from 'react';
import {resolvedSchema as schema, sampleConfig} from 'schema.sample';
import {reqStrPathThrowing, promiseToTask, mergeDeep} from 'rescape-ramda';
import {parentPropsForContainerTask} from 'componentTestHelpers';
import {of} from 'folktale/concurrency/task';
import * as Result from 'folktale/result';
import {loadingCompleteStatus, renderChoicepoint, eMap, renderErrorDefault, renderLoadingDefault} from 'rescape-helpers-component';
import {resolvedRemoteSchemaTask} from './schemaHelpers';
import {addResolvers} from './schema.sample';

const [div] = eMap(['div']);

class App extends Component {
  render() {
    return App.choicepoint(R.merge(this.props, {views: {'error': {}, 'loading': {}, 'success': {}}}));
  }
}

App.choicepoint = renderChoicepoint(
  renderErrorDefault('error'),
  renderLoadingDefault('loading'),
  () => {
    return div({className: 'success'});
  }
);

// This is the views function for our App component.
App.views = props => ({
  error: {
    className: 'error'
  },
  loading: {
    className: 'loading'
  },
  success: {},
  // We'll pretend in the second set of tests that we have a child component,
  // which for convenience is also an App that has a property
  subApp: {
    id: reqStrPathThrowing('data.region.id', props)
  }
});

// Run this apollo query
const query = `query region($regionId: String!) {
      region(id: $regionId) {
            id
      }
}`;
const queries = {
  region: {
    query,
    args: {
      options: ({data: {region}}) => ({
        variables: {
          regionId: region.id
        },
        // Pass through error so we can handle it in the component
        errorPolicy: 'all'
      }),
      props: ({data, ownProps}) => R.merge(
        ownProps,
        {data}
      )
    }
  }
};

const ContainerWithData = graphql(
  gql`${queries.region.query}`,
  queries.region.args
)(App);

// ownProps will override with bad id for testing error
const mapStateToProps = (state, ownProps) => R.merge({data: {region: {id: ownProps.regionId}}}, ownProps);
const mapDispatchToProps = () => ({});
const ContainerClass = connect(mapStateToProps, mapDispatchToProps, R.merge)(ContainerWithData);
const [Container] = eMap([ContainerClass]);

// Find this React component
const componentName = 'App';
// Find this class in the data renderer
const childClassDataName = 'success';
// Find this class in the loading renderer
const childClassLoadingName = 'loading';
// Find this class in the error renderer
const childClassErrorName = 'error';

const errorMaker = parentProps => R.set(R.lensPath(['data', 'region', 'id']), 'foo', parentProps);

// Pretend that there's some parent container that passes the regionId to a view called myContainer, which
// is the container we are testing
const chainedParentPropsTask = parentPropsForContainerTask(
  // Pretend the parent returns the given props asynchronously
  of(Result.Ok({regionId: 'oakland'})),
  props => ({views: {myContainer: {regionId: props.regionId}}}),
  'myContainer'
);

describe('ApolloContainer', () => {
  const {testQuery, testRenderError, testRender} = apolloContainerTests({
    initialState: sampleConfig,
    schema,
    Container,
    chainedParentPropsTask,
    mapStateToProps,
    componentName,
    childClassDataName,
    childClassErrorName,
    childClassLoadingName,
    queryConfig: queries.region,
    errorMaker
  });
  test('testQuery', testQuery);
  test('testRender', testRender);
  test('testRenderError', testRenderError);

  test('makeApolloTestPropsTaskFunction', done => {
    const sampleState = ({data: {regionId: 'oakland'}, views: {aComponent: {stuff: 1}, bComponent: {moreStuff: 2}}});
    const sampleOwnProps = {style: {width: 100}};
    const mapStateToProps = (state, ownProps) => R.merge(state, ownProps);
    const dispatchResults = {
      action1: R.identity,
      action2: R.identity,
      action3: R.identity
    };
    const mapDispatchToProps = (dispatch, ownProps) => dispatchResults;
    // given mapStateToProps, mapDispatchToProps, and mergeProps we get a function back
    // that then takes sample state and ownProps. The result is a merged object based on container methods
    // and sample data. Next apply the apollo query
    const queryObj = {
      query: `
          query region($regionId: String!) {
                region(id: $regionId) {
                    id
                    name
                }
          }
      `,
      args: {
        options: ({data: {regionId}}) => ({
          variables: {
            regionId
          }
        })
      }
    };
    // Make the function with the configuration
    const func = makeApolloTestPropsTaskFunction(schema, sampleConfig, mapStateToProps, mapDispatchToProps, queryObj);
    // Now pretend we're calling it with state and props
    func(sampleState, sampleOwnProps).run().listen({
      // Map the Result, handling Result.Ok success and Result.Error failure
      onResolved: result => result.map(value => {
        expect(value).toEqual(
          R.merge({
            // Expect this data came from Apollo along with the other props that were passed through: style adn views
            data: R.merge(
              loadingCompleteStatus, {
                region: {id: "oakland", name: "Oakland"},
                regionId: 'oakland'
              }),
            style: {width: 100},
            views: {
              aComponent: {stuff: 1},
              bComponent: {moreStuff: 2}
            }
          }, dispatchResults)
        );
        done();
      }).mapError(value => {
        // If any Error values turn up, just throw the errors
        const errors = R.flatten(value.error);
        R.forEach(
          error => {
            console.error(error);
          }
        );
        throw errors[0];
      }),
      onRejected: reject => {
        // If the Task rejects, throw
        throw(reject);
      }
    });
  });

  test('propsFromParentPropsTask', done => {
    propsFromParentPropsTask(
      {foo: 1},
      of(Result.Ok({bar: 1})),
      (initialState, parentContainerSampleProps) => of(R.merge(initialState, parentContainerSampleProps))
    ).run().listen({
      onResolved: value => {
        expect(value).toEqual({
          foo: 1,
          bar: 1
        });
        done();
      }
    });
  });
});


describe('ApolloContainer Remote Integration Test', () => {
  // Test with a remote schema. This is an integration test and requires a server
  // Run the graphql server in the rescape-graphene repo
  const config = {    // This server must be running in order for integration tests to pass
    settings: {
      api: {
        uri: 'http://localhost:8008/graphql'
      },
      // This user must be set up on the server in order for integration tests to pass
      apiAuthorization: {
        username: 'test',
        password: 'testpass'
      }
    }
  };

  const schemaTask = resolvedRemoteSchemaTask(config, addResolvers);
  const {testQuery: testQueryWithRemoteSchema, testRender: testRenderWithRemoteSchema, testRenderError: testRenderErrorWithRemoteSchema} = apolloContainerTests({
    initialState: sampleConfig,
    schema: schemaTask,
    Container,
    chainedParentPropsTask,
    mapStateToProps,
    componentName,
    childClassDataName,
    childClassErrorName,
    childClassLoadingName,
    queryConfig: queries.region,
    errorMaker
  });
  test('testQueryWithRemoteSchema', testQueryWithRemoteSchema);
  test('testRenderWithRemoteSchema', testRenderWithRemoteSchema);
  test('testRenderErrorWithRemoteSchema', testRenderErrorWithRemoteSchema);
});