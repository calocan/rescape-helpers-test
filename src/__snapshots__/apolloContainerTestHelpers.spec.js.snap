// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApolloContainer testMapStateToProps 1`] = `
Object {
  "data": Object {
    "region": Object {
      "id": "oakland",
    },
  },
  "regionId": "oakland",
}
`;

exports[`ApolloContainer testQuery 1`] = `
Object {
  "data": Object {
    "store": Object {
      "__typename": "Store",
      "region": Object {
        "__typename": "Region",
        "id": "oakland",
        Symbol(id): "Region:oakland",
      },
      Symbol(id): "$ROOT_QUERY.store",
    },
    Symbol(id): "ROOT_QUERY",
  },
  "loading": false,
  "networkStatus": 7,
  "stale": false,
}
`;

exports[`ApolloContainer testRender 1`] = `
Object {
  "data": Object {
    "error": undefined,
    "fetchMore": [Function],
    "loading": true,
    "networkStatus": 1,
    "refetch": [Function],
    "startPolling": [Function],
    "stopPolling": [Function],
    "subscribeToMore": [Function],
    "updateQuery": [Function],
    "variables": Object {
      "regionId": "oakland",
    },
  },
  "regionId": "oakland",
}
`;

exports[`ApolloContainer testRender 2`] = `
Object {
  "className": "success",
}
`;

exports[`ApolloContainer testRenderError 1`] = `
Object {
  "data": Object {
    "error": undefined,
    "fetchMore": [Function],
    "loading": true,
    "networkStatus": 1,
    "refetch": [Function],
    "startPolling": [Function],
    "stopPolling": [Function],
    "subscribeToMore": [Function],
    "updateQuery": [Function],
    "variables": Object {
      "regionId": "foo",
    },
  },
  "regionId": "oakland",
}
`;

exports[`ApolloContainer testRenderError 2`] = `
Object {
  "children": "Original Error: Did not find exactly one: { all: [ { id: 'oakland', name: 'Oakland' } ], matching: [] }
Original Trace Error: Did not find exactly one: { all: [ { id: 'oakland', name: 'Oakland' } ], matching: [] }
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/rescape-ramda/lib/throwingFunctions.js:57:11
    at InternalConstructor.mapError (/Users/andy/code/rescape/rescape-helpers-test/node_modules/folktale/result/result.js:291:20)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/rescape-ramda/lib/throwingFunctions.js:54:17
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/ramda/src/internal/_curryN.js:37:27
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/ramda/src/internal/_arity.js:14:19
    at findOneValueByParamsThrowing (/Users/andy/code/rescape/rescape-helpers-test/node_modules/rescape-ramda/lib/throwingFunctions.js:153:10)
    at region (/Users/andy/code/rescape/rescape-helpers-test/src/sampleData.js:87:63)
    at resolveFieldValueOrError (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:531:18)
    at resolveField (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:495:16)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:364:18
    at Array.reduce (<anonymous>)
    at executeFields (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:361:42)
    at collectAndExecuteSubfields (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:772:10)
    at completeObjectValue (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:762:10)
    at completeValue (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:660:12)
    at completeValueWithLocatedError (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:580:21)
    at completeValueCatchingError (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:556:21)
    at resolveField (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:497:10)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:364:18
    at Array.reduce (<anonymous>)
    at executeFields (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:361:42)
    at executeOperation (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:289:122)
    at executeImpl (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:154:14)
    at Object.execute (/Users/andy/code/rescape/rescape-helpers-test/node_modules/graphql/execution/execute.js:131:229)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/apollo-link-schema/src/schemaLink.ts:43:9
    at new Subscription (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:179:34)
    at Observable.subscribe (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:258:14)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/apollo-link-dedup/src/dedupLink.ts:55:41
    at new Subscription (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:179:34)
    at Observable.subscribe (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:258:14)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/QueryManager.ts:1061:66
    at new Promise (<anonymous>)
    at QueryManager.Object.<anonymous>.QueryManager.fetchRequest (/Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/QueryManager.ts:1059:12)
    at QueryManager.Object.<anonymous>.QueryManager.fetchQuery (/Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/QueryManager.ts:375:34)
    at QueryManager.Object.<anonymous>.QueryManager.startQuery (/Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/QueryManager.ts:882:10)
    at ObservableQuery.Object.<anonymous>.ObservableQuery.setUpQuery (/Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/ObservableQuery.ts:590:23)
    at ObservableQuery.Object.<anonymous>.ObservableQuery.onSubscribe (/Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/ObservableQuery.ts:549:43)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/ObservableQuery.ts:96:12
    at new Subscription (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:179:34)
    at ObservableQuery.subscribe (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:258:14)
    at Query._this.startQuerySubscription (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-apollo/react-apollo.umd.js:323:65)
    at Query.Object.<anonymous>.Query.componentDidMount (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-apollo/react-apollo.umd.js:421:18)
    at commitLifeCycles (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:14361:22)
    at commitAllLifeCycles (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:15462:7)
    at Object.invokeGuardedCallback (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:42:10)
    at invokeGuardedCallback (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:187:29)
    at commitRoot (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:15603:7)
    at completeRoot (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16618:34)
    at performWorkOnRoot (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16563:9)
    at performWork (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16482:7)
    at performSyncWork (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16454:3)
    at requestWork (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16354:5)
    at scheduleWork$1 (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16218:11)
    at scheduleRootUpdate (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16785:3)
    at updateContainerAtExpirationTime (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16812:10)
    at updateContainer (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16839:10)
    at ReactRoot.Object.<anonymous>.ReactRoot.render (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:17122:3)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:17262:14
    at unbatchedUpdates (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16679:10)
    at legacyRenderSubtreeIntoContainer (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:17258:5)
    at Object.render (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:17317:12)
    at Object.render (/Users/andy/code/rescape/rescape-helpers-test/node_modules/enzyme-adapter-react-16/build/ReactSixteenAdapter.js:218:50)
    at new ReactWrapper (/Users/andy/code/rescape/rescape-helpers-test/node_modules/enzyme/build/ReactWrapper.js:98:16)
    at mount (/Users/andy/code/rescape/rescape-helpers-test/node_modules/enzyme/build/mount.js:19:10)
    at wrapWithMockGraphqlAndStore (/Users/andy/code/rescape/rescape-helpers-test/src/componentTestHelpers.js:186:28)
    at /Users/andy/code/rescape/rescape-helpers-test/src/apolloContainerTestHelpers.js:237:75
    at Object.onResolved (/Users/andy/code/rescape/rescape-helpers-test/node_modules/folktale/concurrency/task/_task.js:70:13)
    at Object.Resolved (/Users/andy/code/rescape/rescape-helpers-test/node_modules/folktale/concurrency/future/_deferred.js:167:24)
    at InternalConstructor.matchWith (/Users/andy/code/rescape/rescape-helpers-test/node_modules/folktale/adt/union/union.js:67:27)
    at Deferred.listen (/Users/andy/code/rescape/rescape-helpers-test/node_modules/folktale/concurrency/future/_deferred.js:158:17)
    at TaskExecution.listen (/Users/andy/code/rescape/rescape-helpers-test/node_modules/folktale/concurrency/task/_task-execution.js:41:22)
    at Task._computation (/Users/andy/code/rescape/rescape-helpers-test/node_modules/folktale/concurrency/task/_task.js:66:19)
    at Task.run (/Users/andy/code/rescape/rescape-helpers-test/node_modules/folktale/concurrency/task/_task.js:443:28)
    at Object.testRenderError (/Users/andy/code/rescape/rescape-helpers-test/src/apolloContainerTestHelpers.js:242:8)
    at resolve (/Users/andy/code/rescape/rescape-helpers-test/node_modules/jest-jasmine2/build/queue_runner.js:56:12)
    at new Promise (<anonymous>)
    at mapper (/Users/andy/code/rescape/rescape-helpers-test/node_modules/jest-jasmine2/build/queue_runner.js:43:19)
    at promise.then (/Users/andy/code/rescape/rescape-helpers-test/node_modules/jest-jasmine2/build/queue_runner.js:87:41)
Error: GraphQL error: Did not find exactly one: { all: [ { id: 'oakland', name: 'Oakland' } ], matching: [] }
Trace: Error: GraphQL error: Did not find exactly one: { all: [ { id: 'oakland', name: 'Oakland' } ], matching: [] }
    at new ApolloError (/Users/andy/code/rescape/rescape-helpers-test/node_modules/src/errors/ApolloError.ts:56:5)
    at Query._this.getQueryResult (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-apollo/react-apollo.umd.js:367:29)
    at Query.Object.<anonymous>.Query.render (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-apollo/react-apollo.umd.js:465:36)
    at finishClassComponent (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:13193:31)
    at updateClassComponent (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:13155:10)
    at beginWork (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:13824:14)
    at performUnitOfWork (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:15863:12)
    at workLoop (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:15902:24)
    at renderRoot (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:15942:7)
    at performWorkOnRoot (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16560:22)
    at performWork (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16482:7)
    at performSyncWork (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16454:3)
    at requestWork (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16354:5)
    at scheduleWork$1 (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:16218:11)
    at Object.enqueueForceUpdate (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-dom/cjs/react-dom.development.js:11336:5)
    at Query.Object.<anonymous>.Component.forceUpdate (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react/cjs/react.development.js:288:16)
    at Query._this.updateCurrentData (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-apollo/react-apollo.umd.js:358:27)
    at Object.next (/Users/andy/code/rescape/rescape-helpers-test/node_modules/react-apollo/react-apollo.umd.js:329:31)
    at notifySubscription (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:126:18)
    at onNotify (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:161:3)
    at SubscriptionObserver.next (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:215:7)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/ObservableQuery.ts:582:55
    at Array.forEach (<anonymous>)
    at Object.next (/Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/ObservableQuery.ts:582:24)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/QueryManager.ts:598:26
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/QueryManager.ts:1027:11
    at Array.forEach (<anonymous>)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/QueryManager.ts:1026:10
    at Map.forEach (<anonymous>)
    at QueryManager.Object.<anonymous>.QueryManager.broadcastQueries (/Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/QueryManager.ts:1020:18)
    at Object.next (/Users/andy/code/rescape/rescape-helpers-test/node_modules/src/core/QueryManager.ts:1093:18)
    at notifySubscription (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:126:18)
    at onNotify (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:161:3)
    at SubscriptionObserver.next (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:215:7)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/apollo-link-dedup/src/dedupLink.ts:60:43
    at Array.forEach (<anonymous>)
    at Object.next (/Users/andy/code/rescape/rescape-helpers-test/node_modules/apollo-link-dedup/src/dedupLink.ts:60:27)
    at notifySubscription (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:126:18)
    at onNotify (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:161:3)
    at SubscriptionObserver.next (/Users/andy/code/rescape/rescape-helpers-test/node_modules/zen-observable/lib/Observable.js:215:7)
    at /Users/andy/code/rescape/rescape-helpers-test/node_modules/apollo-link-schema/src/schemaLink.ts:56:22",
  "className": "error",
}
`;
